File: scripts/restore-covers-from-csv.py
#!/usr/bin/env python3
"""
Restore bookshelf covers (and ISBNs) from an authoritative StoryGraph CSV.
Prefers Open Library / Google Books covers for the exact ISBN in the CSV.
"""
from __future__ import annotations

import csv
import os
import re
import sys
import time
import urllib.error
import urllib.parse
import urllib.request
from pathlib import Path

ROOT = Path(__file__).resolve().parents[1]
DEFAULT_CSV = Path.home() / "Downloads/ab3fa717133180bd8a861b0e9bfaf78023eb23c30419db3ff86c724993e2f33f.csv"
BOOKS_DIR = ROOT / "src/content/bookshelf"
COVER_DIR = ROOT / "public/covers/bookshelf"

UA = "bookshelf-cover-restore/1.0"


def clean_isbn(raw: str | None) -> str:
    if not raw:
        return ""
    return re.sub(r"[^0-9Xx]", "", str(raw)).upper()


def isbn10_to_13(isbn10: str) -> str | None:
    s = clean_isbn(isbn10)
    if len(s) != 10:
        return None
    core = "978" + s[:9]
    total = sum(int(d) * (1 if i % 2 == 0 else 3) for i, d in enumerate(core))
    check = (10 - (total % 10)) % 10
    return core + str(check)


def to_isbn13(isbn: str | None) -> str | None:
    s = clean_isbn(isbn)
    if len(s) == 13:
        return s
    if len(s) == 10:
        return isbn10_to_13(s)
    return None


def same_isbn(a: str | None, b: str | None) -> bool:
    a13, b13 = to_isbn13(a), to_isbn13(b)
    if a13 and b13 and a13 == b13:
        return True
    return clean_isbn(a) == clean_isbn(b)


def is_valid_isbn(isbn: str | None) -> bool:
    s = clean_isbn(isbn)
    if len(s) == 10:
        return bool(re.match(r"^\d{9}[\dX]$", s))
    if len(s) == 13:
        return bool(re.match(r"^97[89]\d{10}$", s))
    return False


def title_key(s: str | None) -> str:
    t = (s or "").lower()
    t = re.sub(r"\(.*?\)", "", t)
    t = re.sub(r":.*$", "", t)
    t = re.sub(r"[^a-z0-9]+", " ", t)
    t = re.sub(r"\b(the|a|an)\b", " ", t)
    return re.sub(r"\s+", " ", t).strip()


def titles_match(a: str, b: str) -> bool:
    q, c = title_key(a), title_key(b)
    if not q or not c:
        return False
    return q == c or q.startswith(c) or c.startswith(q)


def fetch(url: str, timeout: float = 25.0) -> tuple[int, bytes, str]:
    req = urllib.request.Request(url, headers={"User-Agent": UA})
    try:
        with urllib.request.urlopen(req, timeout=timeout) as res:
            return res.status, res.read(), res.headers.get("Content-Type", "")
    except urllib.error.HTTPError as e:
        return e.code, e.read() if e.fp else b"", e.headers.get("Content-Type", "") if e.headers else ""
    except Exception:
        return 0, b"", ""


def save_cover_from_url(url: str, dest: Path) -> bool:
    status, body, ctype = fetch(url)
    if status != 200:
        return False
    if "image" not in ctype and not url.lower().endswith((".jpg", ".jpeg", ".png", ".webp")):
        # Open Library sometimes omits content-type; accept if body looks like JPEG
        if not (body[:3] == b"\xff\xd8\xff" or body[:8] == b"\x89PNG\r\n\x1a\n"):
            return False
    if len(body) < 1500:
        return False
    dest.write_bytes(body)
    return True


def cover_from_google(isbn: str) -> str | None:
    url = f"https://www.googleapis.com/books/v1/volumes?q=isbn:{isbn}&maxResults=1"
    status, body, _ = fetch(url)
    if status != 200:
        return None
    import json

    try:
        data = json.loads(body.decode("utf-8"))
    except Exception:
        return None
    info = (data.get("items") or [{}])[0].get("volumeInfo", {}).get("imageLinks") or {}
    raw = info.get("extraLarge") or info.get("large") or info.get("medium") or info.get("thumbnail") or info.get("smallThumbnail")
    if not raw:
        return None
    return raw.replace("http://", "https://").replace("&edge=curl", "").replace("zoom=1", "zoom=2")


def cover_from_ol_search(title: str, author: str) -> str | None:
    clean = re.sub(r"\(.*?\)", "", title or "").split(":")[0].strip()
    if not clean:
        return None
    author_last = re.sub(r"[^a-z]", "", (author or "").split(",")[0].strip().split()[-1].lower()) if author else ""
    params = urllib.parse.urlencode({"title": clean, "limit": "10", **({"author": author.split(",")[0].strip()} if author else {})})
    status, body, _ = fetch(f"https://openlibrary.org/search.json?{params}")
    if status != 200:
        return None
    import json

    try:
        docs = json.loads(body.decode("utf-8")).get("docs") or []
    except Exception:
        return None

    def author_ok(d: dict) -> bool:
        if not author_last:
            return True
        return author_last in " ".join(d.get("author_name") or []).lower()

    pick = next((d for d in docs if d.get("cover_i") and author_ok(d) and titles_match(clean, d.get("title") or "")), None)
    if not pick:
        pick = next((d for d in docs if d.get("cover_i") and author_ok(d)), None)
    if not pick:
        return None
    return f"https://covers.openlibrary.org/b/id/{pick['cover_i']}-L.jpg"


def download_cover(isbn: str, dest: Path, title: str, author: str) -> str | None:
    urls: list[str] = []
    if is_valid_isbn(isbn):
        for size in ("L", "M"):
            urls.append(f"https://covers.openlibrary.org/b/isbn/{isbn}-{size}.jpg?default=false")
        as13 = to_isbn13(isbn)
        if as13 and as13 != isbn:
            for size in ("L", "M"):
                urls.append(f"https://covers.openlibrary.org/b/isbn/{as13}-{size}.jpg?default=false")

    for url in urls:
        if save_cover_from_url(url, dest):
            return "openlibrary-isbn"

    if is_valid_isbn(isbn):
        gb = cover_from_google(isbn)
        if gb and save_cover_from_url(gb, dest):
            return "google-isbn"

    search = cover_from_ol_search(title, author)
    if search and save_cover_from_url(search, dest):
        return "openlibrary-search"
    return None


def parse_fm(text: str) -> tuple[str, str] | None:
    if not text.startswith("---"):
        return None
    end = text.find("\n---", 3)
    if end < 0:
        return None
    return text[4:end], text[end + 4 :]


def get_field(block: str, key: str) -> str | None:
    m = re.search(rf"^{re.escape(key)}:\s*(.*)$", block, re.M)
    if not m:
        return None
    return m.group(1).strip().strip('"').strip("'")


def set_field(block: str, key: str, value: str) -> str:
    line = f"{key}: {value}"
    if re.search(rf"^{re.escape(key)}:\s*.*$", block, re.M):
        return re.sub(rf"^{re.escape(key)}:\s*.*$", line, block, count=1, flags=re.M)
    if re.search(r"^author:\s*.*$", block, re.M):
        return re.sub(r"^(author:\s*.*)$", rf"\1\n{line}", block, count=1, flags=re.M)
    return block.rstrip() + "\n" + line + "\n"


def remove_field(block: str, key: str) -> str:
    return re.sub(rf"^{re.escape(key)}:\s*.*\n?", "", block, count=1, flags=re.M)


def yaml_escape(s: str) -> str:
    return '"' + s.replace("\\", "\\\\").replace('"', '\\"') + '"'


def find_csv(csv_by_title: dict, title: str | None, author: str | None):
    key = title_key(title)
    cands = csv_by_title.get(key)
    if not cands:
        cands = csv_by_title.get(title_key((title or "").split(":")[0]))
    if not cands:
        for k, v in csv_by_title.items():
            if key and (key.startswith(k) or k.startswith(key)) and min(len(key), len(k)) > 10:
                cands = v
                break
    if not cands:
        return None
    if len(cands) == 1:
        return cands[0]
    author_l = (author or "").lower()
    for c in cands:
        if author_l and any(p in c["authors"].lower() for p in author_l.split() if len(p) > 2):
            return c
    return cands[0]


def main() -> int:
    csv_path = Path(sys.argv[1]) if len(sys.argv) > 1 else DEFAULT_CSV
    COVER_DIR.mkdir(parents=True, exist_ok=True)

    with csv_path.open(newline="", encoding="utf-8") as f:
        rows = list(csv.DictReader(f))

    csv_by_title: dict[str, list] = {}
    for r in rows:
        isbn = clean_isbn(r.get("ISBN/UID"))
        title = (r.get("Title") or "").strip()
        if not title:
            continue
        key = title_key(title)
        entry = {
            "title": title,
            "authors": r.get("Authors") or "",
            "isbn": isbn if is_valid_isbn(isbn) else None,
        }
        csv_by_title.setdefault(key, []).append(entry)

    files = sorted(BOOKS_DIR.glob("*.md"))
    updated_isbn = replaced = failed = 0

    print(f"CSV: {len(rows)} rows; bookshelf: {len(files)} books")
    print("Restoring covers for ISBN mismatches / missing ISBNs…\n")

    for fp in files:
        text = fp.read_text(encoding="utf-8")
        parsed = parse_fm(text)
        if not parsed:
            continue
        block, body = parsed
        title = get_field(block, "title")
        author = get_field(block, "author")
        md_isbn = clean_isbn(get_field(block, "isbn"))
        cover = get_field(block, "cover")
        slug = fp.stem

        match = find_csv(csv_by_title, title, author)
        if not match or not match["isbn"]:
            continue

        target = match["isbn"]
        if md_isbn and same_isbn(md_isbn, target):
            continue

        block = set_field(block, "isbn", yaml_escape(target))
        updated_isbn += 1

        if cover and cover.startswith("/covers/bookshelf/"):
            dest = ROOT / "public" / cover.lstrip("/")
        else:
            dest = COVER_DIR / f"{slug}.jpg"

        print(f"{slug[:52]:52} {target} ", end="", flush=True)
        tmp = dest.with_suffix(dest.suffix + ".tmp")
        source = download_cover(target, tmp, title or "", author or "")
        if source:
            if dest.exists():
                dest.unlink()
            tmp.rename(dest)
            cover_path = f"/covers/bookshelf/{dest.name}"
            block = set_field(block, "cover", cover_path)
            block = remove_field(block, "coverColor")
            replaced += 1
            print(f"✓ {source}")
        else:
            if tmp.exists():
                tmp.unlink()
            failed += 1
            print("✗ keep existing")

        new_text = f"---\n{block.rstrip()}\n---{body if body.startswith(chr(10)) else chr(10) + body}"
        fp.write_text(new_text, encoding="utf-8")
        time.sleep(0.12)

    print("\nDone.")
    print(f"  ISBNs updated:      {updated_isbn}")
    print(f"  Covers replaced:    {replaced}")
    print(f"  Download failures:  {failed}")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: scripts/restore-covers-from-csv.mjs
#!/usr/bin/env node
/**
 * Restore bookshelf covers (and ISBNs) from an authoritative StoryGraph/Goodreads-style CSV.
 * Prefers Open Library / Google Books covers for the exact ISBN in the CSV over fuzzy title search.
 *
 * Usage:
 *   node scripts/restore-covers-from-csv.mjs [path/to/export.csv]
 */
import { readFile, writeFile, mkdir, access, unlink } from 'node:fs/promises';
import { createReadStream } from 'node:fs';
import path from 'node:path';
import { parse } from 'csv-parse';
import { fileURLToPath } from 'node:url';

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const root = path.resolve(__dirname, '..');
const csvPath =
  process.argv[2] ||
  path.join(
    process.env.HOME || '',
    'Downloads/ab3fa717133180bd8a861b0e9bfaf78023eb23c30419db3ff86c724993e2f33f.csv'
  );
const booksDir = path.join(root, 'src/content/bookshelf');
const coverDir = path.join(root, 'public/covers/bookshelf');

await mkdir(coverDir, { recursive: true });

function cleanIsbn(raw) {
  if (!raw) return '';
  return String(raw).replace(/[^\dXx]/g, '').toUpperCase();
}

function isbn10To13(isbn10) {
  const s = cleanIsbn(isbn10);
  if (s.length !== 10) return null;
  const core = '978' + s.slice(0, 9);
  const sum = [...core].reduce((acc, d, i) => acc + Number(d) * (i % 2 === 0 ? 1 : 3), 0);
  const check = (10 - (sum % 10)) % 10;
  return core + String(check);
}

function toIsbn13(isbn) {
  const s = cleanIsbn(isbn);
  if (s.length === 13) return s;
  if (s.length === 10) return isbn10To13(s);
  return null;
}

function sameIsbn(a, b) {
  const a13 = toIsbn13(a);
  const b13 = toIsbn13(b);
  if (a13 && b13 && a13 === b13) return true;
  return cleanIsbn(a) === cleanIsbn(b);
}

function isValidIsbn(isbn) {
  const s = cleanIsbn(isbn);
  if (s.length === 10) return /^\d{9}[\dX]$/.test(s);
  if (s.length === 13) return /^97[89]\d{10}$/.test(s);
  return false;
}

function titleKey(s) {
  return String(s || '')
    .toLowerCase()
    .normalize('NFKD')
    .replace(/[\u0300-\u036f]/g, '')
    .replace(/\(.*?\)/g, '')
    .replace(/:.*$/, '')
    .replace(/[^a-z0-9]+/g, ' ')
    .replace(/\b(the|a|an)\b/g, ' ')
    .replace(/\s+/g, ' ')
    .trim();
}

function titlesMatch(a, b) {
  const q = titleKey(a);
  const c = titleKey(b);
  if (!q || !c) return false;
  return q === c || q.startsWith(c) || c.startsWith(q);
}

function sleep(ms) {
  return new Promise((r) => setTimeout(r, ms));
}

async function fileExists(p) {
  try {
    await access(p);
    return true;
  } catch {
    return false;
  }
}

async function saveCoverFromUrl(url, dest) {
  const res = await fetch(url, {
    redirect: 'follow',
    headers: { 'User-Agent': 'bookshelf-cover-restore/1.0' },
  });
  if (!res.ok) return false;
  const type = res.headers.get('content-type') || '';
  if (!type.includes('image')) return false;
  const buf = Buffer.from(await res.arrayBuffer());
  if (buf.length < 1500) return false;
  await writeFile(dest, buf);
  return true;
}

async function coverFromGoogleBooks(isbn) {
  try {
    const url = `https://www.googleapis.com/books/v1/volumes?q=isbn:${isbn}&maxResults=1`;
    const res = await fetch(url);
    if (!res.ok) return null;
    const json = await res.json();
    const info = json.items?.[0]?.volumeInfo?.imageLinks;
    if (!info) return null;
    const raw = info.extraLarge || info.large || info.medium || info.thumbnail || info.smallThumbnail;
    if (!raw) return null;
    return raw.replace('http://', 'https://').replace('&edge=curl', '').replace('zoom=1', 'zoom=2');
  } catch {
    return null;
  }
}

async function coverFromOpenLibrarySearch(title, author) {
  const clean = String(title || '')
    .replace(/\(.*?\)/g, '')
    .split(':')[0]
    .trim();
  if (!clean) return null;
  const authorLast = String(author || '')
    .split(',')[0]
    .trim()
    .split(/\s+/)
    .pop()
    ?.toLowerCase()
    .replace(/[^a-z]/g, '');
  const params = new URLSearchParams({ title: clean, limit: '10' });
  if (author) params.set('author', String(author).split(',')[0].trim());
  try {
    const search = await fetch(`https://openlibrary.org/search.json?${params}`);
    if (!search.ok) return null;
    const json = await search.json();
    const docs = json.docs || [];
    const authorOk = (d) => {
      if (!authorLast) return true;
      return (d.author_name || []).join(' ').toLowerCase().includes(authorLast);
    };
    const pick =
      docs.find((d) => d.cover_i && authorOk(d) && titlesMatch(clean, d.title)) ||
      docs.find((d) => d.cover_i && authorOk(d));
    if (!pick?.cover_i) return null;
    return `https://covers.openlibrary.org/b/id/${pick.cover_i}-L.jpg`;
  } catch {
    return null;
  }
}

/** Force-download cover for an ISBN; overwrite existing file. */
async function downloadCoverForIsbn(isbn, dest, title, author) {
  const urls = [];
  if (isbn && isValidIsbn(isbn)) {
    urls.push(
      `https://covers.openlibrary.org/b/isbn/${isbn}-L.jpg?default=false`,
      `https://covers.openlibrary.org/b/isbn/${isbn}-M.jpg?default=false`
    );
    const as13 = toIsbn13(isbn);
    if (as13 && as13 !== isbn) {
      urls.push(
        `https://covers.openlibrary.org/b/isbn/${as13}-L.jpg?default=false`,
        `https://covers.openlibrary.org/b/isbn/${as13}-M.jpg?default=false`
      );
    }
  }

  for (const url of urls) {
    try {
      if (await saveCoverFromUrl(url, dest)) return { ok: true, source: 'openlibrary-isbn' };
    } catch {
      // next
    }
  }

  if (isbn && isValidIsbn(isbn)) {
    const gb = await coverFromGoogleBooks(isbn);
    if (gb) {
      try {
        if (await saveCoverFromUrl(gb, dest)) return { ok: true, source: 'google-isbn' };
      } catch {
        // fall through
      }
    }
  }

  const searchUrl = await coverFromOpenLibrarySearch(title, author);
  if (searchUrl) {
    try {
      if (await saveCoverFromUrl(searchUrl, dest)) return { ok: true, source: 'openlibrary-search' };
    } catch {
      // give up
    }
  }
  return { ok: false, source: null };
}

function parseFrontmatter(text) {
  if (!text.startsWith('---\n') && !text.startsWith('---\r\n')) return null;
  const end = text.indexOf('\n---', 3);
  if (end < 0) return null;
  return {
    block: text.slice(4, end),
    body: text.slice(end + 4),
    full: text,
  };
}

function getFmField(block, key) {
  const re = new RegExp(`^${key}:\\s*(.*)$`, 'm');
  const m = block.match(re);
  if (!m) return null;
  return m[1].trim().replace(/^["']|["']$/g, '');
}

function setFmField(block, key, value) {
  const re = new RegExp(`^${key}:\\s*.*$`, 'm');
  const line = `${key}: ${value}`;
  if (re.test(block)) return block.replace(re, line);
  // insert after author if present, else at end
  if (/^author:/m.test(block)) {
    return block.replace(/^(author:\s*.*)$/m, `$1\n${line}`);
  }
  return `${block.trimEnd()}\n${line}\n`;
}

function removeFmField(block, key) {
  return block.replace(new RegExp(`^${key}:\\s*.*\\n?`, 'm'), '');
}

function yamlEscape(s) {
  return `"${String(s).replace(/\\/g, '\\\\').replace(/"/g, '\\"')}"`;
}

// Load CSV
const rows = [];
await new Promise((resolve, reject) => {
  createReadStream(csvPath)
    .pipe(parse({ columns: true, relax_quotes: true, relax_column_count: true }))
    .on('data', (row) => rows.push(row))
    .on('error', reject)
    .on('end', resolve);
});

const csvByTitle = new Map();
for (const r of rows) {
  const isbn = cleanIsbn(r['ISBN/UID']);
  const title = String(r.Title || '').trim();
  if (!title) continue;
  const key = titleKey(title);
  const entry = {
    title,
    authors: String(r.Authors || ''),
    isbn: isValidIsbn(isbn) ? isbn : null,
  };
  if (!csvByTitle.has(key)) csvByTitle.set(key, []);
  csvByTitle.get(key).push(entry);
}

function findCsv(title, author) {
  const key = titleKey(title);
  let cands = csvByTitle.get(key);
  if (!cands) {
    const short = titleKey(String(title || '').split(':')[0]);
    cands = csvByTitle.get(short);
  }
  if (!cands) {
    for (const [k, v] of csvByTitle) {
      if (key && (key.startsWith(k) || k.startsWith(key)) && Math.min(key.length, k.length) > 10) {
        cands = v;
        break;
      }
    }
  }
  if (!cands?.length) return null;
  if (cands.length === 1) return cands[0];
  const authorL = String(author || '').toLowerCase();
  for (const c of cands) {
    if (authorL && authorL.split(/\s+/).some((p) => p.length > 2 && c.authors.toLowerCase().includes(p))) {
      return c;
    }
  }
  return cands[0];
}

const { readdir } = await import('node:fs/promises');
const files = (await readdir(booksDir)).filter((f) => f.endsWith('.md')).sort();

let updatedIsbn = 0;
let replacedCovers = 0;
let keptCovers = 0;
let failed = 0;
const report = [];

console.log(`CSV: ${rows.length} rows; bookshelf: ${files.length} books`);
console.log(`Restoring covers preferring exact CSV ISBNs…\n`);

for (const file of files) {
  const fp = path.join(booksDir, file);
  const text = await readFile(fp, 'utf8');
  const parsed = parseFrontmatter(text);
  if (!parsed) continue;

  let block = parsed.block;
  const title = getFmField(block, 'title');
  const author = getFmField(block, 'author');
  const mdIsbn = cleanIsbn(getFmField(block, 'isbn'));
  const cover = getFmField(block, 'cover');
  const slug = path.basename(file, '.md');

  const csv = findCsv(title, author);
  if (!csv?.isbn) {
    continue;
  }

  const targetIsbn = csv.isbn;
  const isbnChanged = !mdIsbn || !sameIsbn(mdIsbn, targetIsbn);
  // Only rewrite covers when the preferred edition ISBN differs or was missing
  // (StoryGraph title-search backfill often filled those with the wrong edition).
  if (!isbnChanged) continue;

  block = setFmField(block, 'isbn', yamlEscape(targetIsbn));
  updatedIsbn += 1;

  const dest = cover?.startsWith('/covers/bookshelf/')
    ? path.join(root, 'public', cover)
    : path.join(coverDir, `${slug}.jpg`);

  process.stdout.write(`${slug.slice(0, 52).padEnd(52)} ${targetIsbn} `);

  // Write to temp then replace so a failed download doesn't wipe a good cover
  const tmp = dest + '.tmp';
  const result = await downloadCoverForIsbn(targetIsbn, tmp, title, author);
  if (result.ok) {
    // replace
    try {
      if (await fileExists(dest)) await unlink(dest);
    } catch {
      // ignore
    }
    const { rename } = await import('node:fs/promises');
    await rename(tmp, dest);
    const coverPath = `/covers/bookshelf/${path.basename(dest)}`;
    block = setFmField(block, 'cover', coverPath);
    // drop stale dominant color so enrich can recompute later
    block = removeFmField(block, 'coverColor');
    replacedCovers += 1;
    console.log(`✓ ${result.source}`);
    report.push({ file, title, isbn: targetIsbn, source: result.source, isbnChanged });
  } else {
    try {
      if (await fileExists(tmp)) await unlink(tmp);
    } catch {
      // ignore
    }
    failed += 1;
    keptCovers += 1;
    console.log('✗ keep existing');
    report.push({ file, title, isbn: targetIsbn, source: 'failed', isbnChanged });
  }

  if (block !== parsed.block) {
    await writeFile(fp, `---\n${block.trimEnd()}\n---${parsed.body.startsWith('\n') ? '' : '\n'}${parsed.body}`);
  }

  await sleep(120);
}

console.log(`\nDone.`);
console.log(`  ISBNs updated:     ${updatedIsbn}`);
console.log(`  Covers replaced:   ${replacedCovers}`);
console.log(`  Covers kept (fail): ${keptCovers}`);
console.log(`  Download failures: ${failed}`);
